pkg/lifecyclehooks/lyfecyclehooks.go

```diff
@@ -0,0 +1,207 @@
+		By("Running a workload on the machine")
+		// Run a pod on this machine
+		workloadMemRequest := resource.MustParse("100000000")
+		workload = newWorkLoad(int32(expectedReplicas), workloadMemRequest, targetedNodeLabel)
+		Expect(client.Create(ctx, workload)).Should(Succeed())
+		Expect(err).ToNot(HaveOccurred())
+		By("Waiting for job pod to start running on machine.")
+		Eventually(func() bool {
+			jobPodList, err := framework.GetPods(client, map[string]string{lifecycleHooksTestLabel: ""})
+			Expect(err).ToNot(HaveOccurred())
+			pod = &jobPodList.Items[0]
+			return len(jobPodList.Items) == int(expectedReplicas) && pod.Status.Phase == corev1.PodRunning
+		}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+	})
+
+	AfterEach(func() {
+		By("Deleting the machineset")
+		cascadeDelete := metav1.DeletePropagationForeground
+		Expect(client.Delete(ctx, machineSet, &runtimeclient.DeleteOptions{
+			PropagationPolicy: &cascadeDelete,
+		})).To(Succeed())
+		By("Deleting workload job")
+		Expect(client.Delete(ctx, workload, &runtimeclient.DeleteOptions{
+			PropagationPolicy: &cascadeDelete,
+		}))
+	})
+
+	It("pause lifecycle actions when present", func() {
+		By("Setting lifecycle hooks on the machine")
+		// Create a pre-drain hook for the machine
+		predrainHook := &v1beta1.LifecycleHook{
+			Name:  "cluster-api-actuator-pkg/pre-drainHook",
+			Owner: "cluster-api-actuator-pkg",
+		}
+		machines, err := framework.GetMachinesFromMachineSet(client, machineSet)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machines).To(HaveLen(1))
+		machine = machines[0]
+		machine.Spec.LifecycleHooks.PreDrain = []v1beta1.LifecycleHook{*predrainHook}
+		err = client.Update(context.Background(), machine)
+		Expect(err).ToNot(HaveOccurred())
+		//Create a pre-terminate hook for the machine
+		preterminateHook := &v1beta1.LifecycleHook{
+			Name:  "cluster-api-actuator-pkg/pre-terminateHook",
+			Owner: "cluster-api-actuator-pkg",
+		}
+		Eventually(func() bool {
+			key := types.NamespacedName{Namespace: machine.Namespace, Name: machine.Name}
+			err = client.Get(ctx, key, machine)
+			Expect(err).ToNot(HaveOccurred())
+			machine.Spec.LifecycleHooks.PreTerminate = []v1beta1.LifecycleHook{*preterminateHook}
+			err = client.Update(context.Background(), machine)
+			if apierrors.IsConflict(err) {
+				return false
+			} else {
+				Expect(err).ToNot(HaveOccurred())
+			}
+			return true
+		}, framework.WaitShort, pollingInterval).Should(BeTrue())
+		By("Deleting the machine")
+		// Delete the machine by scaling down the machineset to zero
+		framework.ScaleMachineSet(machineSet.Name, 0)
+		By("Checking that workload pod is running on machine")
+		// pre-drain hook should prevent pod from being evicted
+		key := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+		err = client.Get(ctx, key, pod)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(pod.Status.Phase).To(Equal(corev1.PodRunning))
```

Comment:
can we have an `Eventually()` block here to make sure that pod was not stopped for at least a couple of minutes?